samza-sql/src/main/java/org/apache/samza/sql/translator/JoinTranslator.java

```diff
@@ -0,0 +1,319 @@
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.sql.SqlExplainFormat;
+import org.apache.calcite.sql.SqlExplainLevel;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.commons.lang.Validate;
+import org.apache.samza.SamzaException;
+import org.apache.samza.operators.KV;
+import org.apache.samza.operators.MessageStream;
+import org.apache.samza.operators.functions.StreamTableJoinFunction;
+import org.apache.samza.serializers.KVSerde;
+import org.apache.samza.sql.data.SamzaSqlCompositeKey;
+import org.apache.samza.sql.data.SamzaSqlRelMessage;
+import org.apache.samza.sql.interfaces.SourceResolver;
+import org.apache.samza.sql.serializers.SamzaSqlCompositeKeySerdeFactory;
+import org.apache.samza.sql.serializers.SamzaSqlRelMessageSerdeFactory;
+import org.apache.samza.storage.kv.RocksDbTableDescriptor;
+import org.apache.samza.table.Table;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.samza.sql.data.SamzaSqlCompositeKey.*;
+import static org.apache.samza.sql.serializers.SamzaSqlCompositeKeySerdeFactory.*;
+import static org.apache.samza.sql.serializers.SamzaSqlRelMessageSerdeFactory.*;
+
+
+/**
+ * Translator to translate the LogicalJoin node in the relational graph to the corresponding StreamGraph
+ * implementation.
+ * Join is supported with the following caveats:
+ *   1. Only stream-table joins are supported. No stream-stream joins.
+ *   2. Only Equi-joins are supported. No theta-joins.
+ *   3. Inner joins, Left and Right outer joins are supported. No cross joins, full outer joins or natural joins.
+ *   4. Compound join condition with only AND operator is supported. AND operator with a literal is not supported. No
+ *      support for OR operator or any other operator in the join condition.
+ *
+ * It is assumed that the stream denoted as 'table' is already partitioned by the key(s) specified in the join
+ * condition. We do not repartition the table. But we always repartition the stream by the key(s) specified
+ * in the join condition.
+ */
+public class JoinTranslator {
+
+  private static final Logger log = LoggerFactory.getLogger(JoinTranslator.class);
+  private int joinId;
+  private SourceResolver sourceResolver;
+
+  public JoinTranslator(int joinId, SourceResolver sourceResolver) {
+    this.joinId = joinId;
+    this.sourceResolver = sourceResolver;
+  }
+
+  public void translate(final LogicalJoin join, final TranslatorContext context) {
+
+    // Do the validation of join query
+    validateJoinQuery(join);
+
+    boolean isTablePosOnRight = isTable(join.getRight());
+    List<Integer> streamIds = new LinkedList<>();
+    List<Integer> tableIds = new LinkedList<>();
+
+    // Get the stream and table indices corresponding to the fields given in the join condition.
+    getStreamAndTableKeyIds(((RexCall) join.getCondition()).getOperands(), join, isTablePosOnRight, streamIds, tableIds);
+
+    MessageStream<SamzaSqlRelMessage> inputTable =
```

Comment:
consider moving  lines 94 to 112 to its own method.

This is the logic for loading the table to RocksDb; but that can change depending of the type of store we use in the future. Could be nice to keep this part of the code cleaner of that detail.